### PR TITLE
Integration Test for CPAN - Starting

### PR DESCRIPTION
**Pull request by allenhsieh**
This pull request makes the following changes:
* start of my first integration test
* travis.yml changed to use openjdk8 explicitly (@DarthHater found that TravisCI defaults to openjdk11 now)

tests run with `mvn clean verify` and pass

### CODE REVIEW

File: nexus-repository-cpan/pom.xml [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+                        <phase>verify</phase>
+                        <goals>
+                            <goal>install</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+</project>
```
Comment (aornatovskyy) [inline nexus-repository-cpan/pom.xml]: missed next line

File: nexus-repository-cpan-it/src/test/java/org/sonatype/nexus/repository/cpan/internal/CpanProxyIT.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.sonatype.goodies.httpfixture.server.fluent.Behaviours.content;
+import static org.sonatype.goodies.httpfixture.server.fluent.Behaviours.error;
+import static org.sonatype.nexus.testsuite.testsupport.FormatClientSupport.status;
+
+public class CpanProxyIT
+    extends CpanITSupport
+{
+  private static final String TEST_PATH = "authors/id/E/EH/EHUELS/Test-Dependencies-0.24.tar.gz";
```
Comment (aornatovskyy) [inline nexus-repository-cpan-it/src/test/java/org/sonatype/nexus/repository/cpan/internal/CpanProxyIT.java]: Name of the variable has general meaning. Usually, we give more specific names relates to its value.

File: nexus-repository-cpan-it/pom.xml [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+                        <goals>
+                            <goal>integration-test</goal>
+                            <goal>verify</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+</project>
```
Comment (aornatovskyy) [inline nexus-repository-cpan-it/pom.xml]: missed new line

File: .travis.yml [hunk 4]
```diff
@@ -1 +1,4 @@
 language: java
+
```
Comment (aornatovskyy) [inline .travis.yml]: are we going to use Travis?

File: pom.xml [hunk 5]
```diff
@@ -1,96 +1,92 @@
...
-      <plugin>
-        <groupId>org.apache.maven.plugins</groupId>
-        <artifactId>maven-surefire-plugin</artifactId>
-        <version>2.20</version>
-      </plugin>
-
-    </plugins>
-  </build>
+  <artifactId>cpan-parent</artifactId>
+  <version>0.0.0</version>
```
[the remark this answers is not in the document]
Comment (DarthHater) [reply pom.xml]: We did this on purpose, because CircleCI is going to rev it to 0.0.1 as it releases it.

File: .travis.yml (hunk 4 above)
Comment (DarthHater) [reply .travis.yml]: We kept it on Travis at the moment so as to not include too much in the PR.

File: nexus-repository-cpan-it/src/test/java/org/sonatype/nexus/repository/cpan/internal/CpanProxyIT.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+import org.ops4j.pax.exam.Configuration;
+import org.ops4j.pax.exam.Option;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.sonatype.goodies.httpfixture.server.fluent.Behaviours.content;
+import static org.sonatype.goodies.httpfixture.server.fluent.Behaviours.error;
+import static org.sonatype.nexus.testsuite.testsupport.FormatClientSupport.status;
+
+public class CpanProxyIT
```
[the remark this answers is not in the document]
Comment (DarthHater) [reply nexus-repository-cpan-it/src/test/java/org/sonatype/nexus/repository/cpan/internal/CpanProxyIT.java]: Those are all great ideas, but we'd probably add them after this PR, this was mainly to just setup the infrastructure for additional tests (not write them all)

File: nexus-repository-cpan-it/pom.xml (hunk 3 above)
Comment (DarthHater) [reply nexus-repository-cpan-it/pom.xml]: In: ea67e7d

File: nexus-repository-cpan/pom.xml (hunk 1 above)
Comment (DarthHater) [reply nexus-repository-cpan/pom.xml]: In: ea67e7d

File: pom.xml [hunk 7]
```diff
@@ -1,96 +1,92 @@
...
+                <gpgArgument>--pinentry-mode</gpgArgument>
+                <gpgArgument>loopback</gpgArgument>
+              </gpgArguments>
+            </configuration>
+          </plugin>
+        </plugins>
+      </build>
+    </profile>
+  </profiles>
 </project>
```
[the remark this answers is not in the document]
Comment (DarthHater) [reply pom.xml]: In: ea67e7d

File: nexus-repository-cpan-it/src/test/java/org/sonatype/nexus/repository/cpan/internal/CpanProxyIT.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+      proxyRepo = repos.createCpanProxy("cpan-test-proxy-online", server.getUrl().toExternalForm());
+      proxyClient = cpanClient(proxyRepo);
+      proxyClient.get(TEST_PATH);
+    }
+    finally {
+      server.stop();
+    }
+    assertThat(status(proxyClient.get(TEST_PATH)), is(200));
+  }
+}
```
[the remark this answers is not in the document]
Comment (DarthHater) [reply nexus-repository-cpan-it/src/test/java/org/sonatype/nexus/repository/cpan/internal/CpanProxyIT.java]: In: ea67e7d

File: nexus-repository-cpan-it/src/test/java/org/sonatype/nexus/repository/cpan/internal/fixtures/RepositoryRuleCpan.groovy [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+import org.sonatype.nexus.testsuite.testsupport.fixtures.RepositoryRule
+
+class RepositoryRuleCpan
+    extends RepositoryRule
+    implements CpanRepoRecipes
+{
+  RepositoryRuleCpan(final Provider<RepositoryManager> repositoryManagerProvider) {
+    super(repositoryManagerProvider)
+  }
+}
```
[the remark this answers is not in the document]
Comment (DarthHater) [reply nexus-repository-cpan-it/src/test/java/org/sonatype/nexus/repository/cpan/internal/fixtures/RepositoryRuleCpan.groovy]: In: ea67e7d

File: nexus-repository-cpan-it/src/test/java/org/sonatype/nexus/repository/cpan/internal/CpanProxyIT.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    }
+  }
+
+  @Test
+  public void retrieveTarGzWhenRemoteOnline() throws Exception {
+    Server server = Server.withPort(0).serve("/*")
+        .withBehaviours(Behaviours.file(testData.resolveFile("Test-Dependencies-0.24.tar.gz")))
+        .start();
+    try {
+      proxyRepo = repos.createCpanProxy("cpan-test-proxy-online", server.getUrl().toExternalForm());
```
[the remark this answers is not in the document]
Comment (DarthHater) [reply nexus-repository-cpan-it/src/test/java/org/sonatype/nexus/repository/cpan/internal/CpanProxyIT.java]: In: 912b704

File: nexus-repository-cpan-it/src/test/java/org/sonatype/nexus/repository/cpan/internal/CpanProxyIT.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+        nexusFeature("org.sonatype.nexus.plugins", "nexus-repository-cpan")
+    );
+  }
+
+  @Test
+  public void unresponsiveRemoteProduces404() throws Exception {
+    Server server = Server.withPort(0).serve("/*")
+        .withBehaviours(error(HttpStatus.NOT_FOUND))
+        .start();
+    try {
```
[the remark this answers is not in the document]
Comment (DarthHater) [reply nexus-repository-cpan-it/src/test/java/org/sonatype/nexus/repository/cpan/internal/CpanProxyIT.java]: In: 912b704

File: nexus-repository-cpan-it/src/test/java/org/sonatype/nexus/repository/cpan/internal/CpanProxyIT.java (hunk 2 above)
Comment (DarthHater) [reply nexus-repository-cpan-it/src/test/java/org/sonatype/nexus/repository/cpan/internal/CpanProxyIT.java]: In: 5f4f849

**Comment by DarthHater** [comment]
@aornatovskyy @bhamail I think this PR is ready, would love for one of you fine people to JAH BLESS it for @allenhsieh :)

File: pom.xml [hunk 12]
```diff
@@ -1,96 +1,92 @@
...
+  <properties>
+    <nxrm-version>3.19.0-SNAPSHOT</nxrm-version>
+  </properties>
+  <repositories>
+    <!-- ensure we can find the parent pom when starting from an empty local .m2 repository -->
+    <repository>
+      <id>rso-snapshots</id>
+      <url>https://repository.sonatype.org/content/repositories/snapshots</url>
+    </repository>
+  </repositories>
```
Comment (bhamail) [inline pom.xml]: Optional enhancement (already included in [nexus-format-archetype](https://github.com/sonatype-nexus-community/nexus-format-archetype)): Detect and require java 8. Could do so by adding the snippet below:

```
<build>
  <plugins>
      <!--
      Ensure use of Java 8, as required by NXRM
      -->
      <plugin>
        <artifactId>maven-enforcer-plugin</artifactId>
        <version>1.4.1</version>
        <executions>
          <execution>
            <id>enforce-java</id>
            <goals>
              <goal>enforce</goal>
            </goals>
            <configuration>
              <rules>
                <requireJavaVersion>
                  <version>[1.8,1.9)</version>
                </requireJavaVersion>
              </rules>
            </configuration>
          </execution>
        </executions>
      </plugin>
  </plugins>
</build>
```

Comment (DarthHater) [reply pom.xml]: In: a277059

**Pull request merged**